---
title: "DEArmarkdown"
output: html_document
date: "2024-05-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# prerequisities

### install DESeq2 and EdgeR packages
```{r, eval=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("DESeq2")

if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("edgeR"
```
### set working directory
```{r}
setwd("Z:/FOR-271/RNA_seq_DEA")
getwd()
```
# Count normalization
### read data files
```{r}
#load edgeR
library(edgeR)
Coinfection.targets<-read.delim("./data/fileDesc.txt")
#change names of dataframe
rownames(Coinfection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
#read text files
Coinfection.orig <- readDGE(Coinfection.targets, header=F)
#check dimensions of dataset
dim(Coinfection.orig)
#check first 6 rows
head(Coinfection.orig)
```
dataframe count
```{r}
 Coinfection.rawCount <- Coinfection.orig$count
dim(Coinfection.rawCount)
head(Coinfection.rawCount)
```
### build meta data
```{r}
#define sampletype
sampletype <- factor(c(rep("Ha",3), rep("Ctr", 3)))
#build metadata frame
meta <- data.frame(sampletype, row.names = colnames(Coinfection.orig$count))
#check col names
colnames(Coinfection.orig$count)
#check row names
rownames(meta)
  #check that sample names match
all(colnames(Coinfection.orig$count) %in% rownames(meta))
```
### create DESeqDataset object
load deseq
```{r, message=FALSE}
library(DESeq2)
```
create dataset
```{r}
dds <- DESeqDataSetFromMatrix(Coinfection.orig, colData = meta, design = ~ sampletype)
#view dds
head(counts(dds))
```
### generate size factors
```{r}
dds <- estimateSizeFactors(dds)
#view normalization factors
sizeFactors(dds)
```
### generate normalized counts matrix from dds
```{r}
normalized_counts <- counts(dds, normalized=TRUE)
#save them
write.csv(normalized_counts, file="./results/coinfection_normalized_counts_DESeq2.csv")
```

# Sample-level quality control
## Principal Component Analysis (PCA)
```{r}
#transform counts for data visualization
rld <- rlog(dds, blind=TRUE)
#run PCA
plotPCA(rld, intgroup="sampletype")
#save plot
pdf("./results/PlotPCA_dds.pdf")
plotPCA(rld, intgroup="sampletype")
dev.off()
```
## Hierarhical clustering heatmap
```{r}
#extract rlog matrix
rld_mat <- assay(rld)
#compute pairwise correlation
rld_cor <- cor(rld_mat) 
#check output
head(rld_cor)
```
check meta
```{r}
head(meta)
```
install heatmap
```{r, eval=FALSE}
install.packages("pheatmap", repos = "http://cran.us.r-project.org")
```
plot heatmap
```{r}
library(pheatmap)
pheatmap(rld_cor, annotation = meta)
#change colors
heat.colors <- RColorBrewer::brewer.pal(6, "Blues")
pheatmap(rld_cor, annotation = meta, color = heat.colors, border_color=NA, fontsize = 10, 
         fontsize_row = 10, height=20)
#save heatmap
pdf("./results/PlotHeatmap_dds.pdf")
heat.colors <- RColorBrewer::brewer.pal(6, "Blues")
pheatmap(rld_cor, annotation = meta, color = heat.colors, border_color=NA, fontsize = 10, 
         fontsize_row = 10, height=20)
```

# Differential expression analysis (DEA)
setup edgeR
```{r}
#setwd
setwd("Z:/FOR-271/RNA_seq_DEA")
getwd()
#load edgeR
library(edgeR)
options(digits=3)
#tell R where data is
infection.targets<-read.delim("./data/fileDesc.txt")
```
check targets
```{r}
infection.targets
```
change row names and check targets
```{r}
rownames(infection.targets)<-c("Ha1","Ha2","Ha3","Ctr1","Ctr2","Ctr3")
infection.targets
```
read and merge files containing gene expression to make DGEList
```{r}
infection <- readDGE(infection.targets, header=F)
#check dimension of DGEList
dim(infection)
head(infection)
```
get raw mapped count before filtering
```{r}
infection.rawCount <- infection$count
head(infection.rawCount)
```
install ggplot2
```{r, message=FALSE}
install.packages("ggplot2", repos = "http://cran.us.r-project.org")
library(ggplot2)
```
plot histogram of counts for single samples Ha1
```{r}
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
#export png
png("./results/count distribution.png", res=300, height=1800, width=1800)
ggplot(infection.rawCount) +
  geom_histogram(aes(x = Ha1), stat = "bin", bins = 200) +
  xlab("Raw expression counts") +
  ylab("Number of genes")
dev.off()
```
export raw count table
```{r}
write.csv(infection.rawCount, file="./results/infection.rawCounts.csv")
  #get counts per million
infection.normCPM <- cpm(calcNormFactors(infection))
dim(infection.normCPM)
head(infection.normCPM)
write.csv(infection.normCPM, file="./results/infection.normCPM.csv")
```
filter genes
```{r}
infection.filtered <- rowSums(cpm(infection)>1) >=3
table(infection.filtered)
```
library size before filtering
```{r}
infection$samples$lib.size
```
cover og data with filtered data
```{r}
Infection <- infection[infection.filtered,]
#size after filtering
colSums(Infection$counts)
dim(Infection)
```
update filtered libraries
```{r}
Infection$samples$lib.size <- colSums(Infection$counts)
Infection$samples
```
normalization with TMM
```{r}
Infection = calcNormFactors(Infection)
  #libs after normalization
Infection$samples
  #counts per million after filter (TMM normalized)
Infection.filtered.normCPM <-cpm(calcNormFactors(Infection))
  #export TMM normalized count table
write.csv(Infection.filtered.normCPM, file="./results/Infection.filtered.normCPM.csv")
```
## Differential gene expression analysis
### treatment factor and eperimental design description
```{r}
group<-factor(c('Ha','Ha','Ha',"Ctr","Ctr","Ctr"))
Infection.design <- model.matrix(~group)   
rownames(Infection.design)<-colnames(Infection$counts)
Infection.design
```
plot
```{r}
plotMDS(Infection, main="MDS plot of RNA-Seq", labels=colnames(Infection$counts))
  #save plot
png("./results/plotMDS.Infection.png", res=300, height=1800, width=1800)
plotMDS(Infection, main="MDS plot of Infection RNA-Seq", labels=colnames(Infection$counts))
dev.off()
```
### Estimate the common dispersion, trended dispersion, tagwwise dispersion
```{r}
#estimate common dispersion
Infection <- estimateGLMCommonDisp(Infection, Infection.design)
#estimate trended dispersion
Infection <- estimateGLMTrendedDisp(Infection, Infection.design)
#estimate tagwise dispersion
Infection <- estimateGLMTagwiseDisp(Infection, Infection.design)
#plot mean variance relationship and BCV-mean logCPM relationship
plotMeanVar(Infection, show.tagwise.vars=T,NBline=T)
plotBCV(Infection)
```
### Fit DGEList and design matrix to genewise GLM
```{r}
Infection.fit <- glmFit(Infection, Infection.design)
colnames(Infection.fit)
#likelihood for different gene responses
lrt.Ha_vs_Ctr <- glmLRT(Infection.fit, coef=2)
t1<-topTags(lrt.Ha_vs_Ctr, n=nrow(Infection))
head(t1$table)
```
number of differentially expressed genes
```{r}
summary(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))
```
upregulated genes
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC > 0))
lrt.Ha_vs_Ctr_UP <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC > 0)
```
downregulated genes
```{r}
nrow(subset(topTags(lrt.Ha_vs_Ctr, n=586)$table,  logFC < 0))
lrt.Ha_vs_Ctr_DW <- subset(topTags(lrt.Ha_vs_Ctr, n=586)$table, logFC < 0)
```
differentially expressed transcriptsâ€™ tag
```{r}
DEtags.lrt.Ha_vs_Ctr <- rownames(Infection)[as.logical(decideTests(lrt.Ha_vs_Ctr, adjust.method="BH", p.value=0.05))]
#export up and downregulated transcripts
write.csv(lrt.Ha_vs_Ctr_UP, file="./results/lrt.Ha_vs_Ctr_UP.csv")
write.csv(lrt.Ha_vs_Ctr_DW, file="./results/lrt.Ha_vs_Ctr_DW.csv")
```
### plot
```{r}
 #label genes as grey
Infection.colHavsCtr = rep('grey55', nrow(Infection))
  #color DE transcripts
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC >0 ] <- "red"
Infection.colHavsCtr[lrt.Ha_vs_Ctr$table$PValue < 0.05 & lrt.Ha_vs_Ctr$table$logFC <0 ] <- "blue"
  #plot
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, de.tags=DEtags.lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", pch=19, cex=0.4, smearWidth=0.5, panel.first=grid(), smooth.scatter=FALSE, ylim=c(-7,7), yaxs="i")
#abline
abline(h=c(-1,1),col="dodgerblue")
#plot with changed colors
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
#abline
abline(h=c(-1,1),col="dodgerblue")
#plot smear plot with changed colors and save as png
png("./results/plotSmear.InfectionRNAseq.png", res=300, height=1800, width=1800)
par(omi=c(0.1,0.1,0.1,0.1), las=1, cex=0.5, mgp=c(3,1,0), cex.main=1.8, cex.lab=1.4, cex.axis=1.4)
plotSmear(lrt.Ha_vs_Ctr, xlab="log-counts per million (logCPM)", ylab="log2-fold change (log2FC)", main="Ha infection compared to Control", smearWidth=0.5, pch=21, cex=0.4, deCol="red", col=Infection.colHavsCtr, ylim=c(-7,7), yaxs="i")
abline(h=c(-1,1),col="dodgerblue")
dev.off()

```


